Improper integral does not evaluate #10445

shubhamtibra opened the pull request:
Modified the code to compute the limit of antiderivative when upper-limit is Infinity. #10445

shubhamtibra committed:
Improper integral does not evaluate #10445

Modified the code to compute the limit of antiderivative when upper-limit is Infinity.

jksuom commented on sympy/integrals/integrals.py:
```diff
@@ -520,8 +521,14 @@ def try_meijerg(function, xab):
                         function = antideriv._eval_interval(x, a, b)
                         function = Poly(function, *gens)
                     elif isinstance(antideriv, Add):
```
I think this alternative should be entirely removed. The whole ``antideriv`` could be passed to ``_eval_interval`` as is done after the next ``else:`` statement.

shubhamtibra commented on sympy/integrals/integrals.py:
```diff
@@ -520,8 +521,14 @@ def try_meijerg(function, xab):
+                                Add.make_args(antideriv)])
                     else:
                         try:
                             function = antideriv._eval_interval(x, a, b)
```
@jksuom Are you talking about this part? Should we remove the `try` then?

shubhamtibra committed:
Removed Add.make_args

shubhamtibra commented:
One test is failing [link](https://github.com/sympy/sympy/blob/master/sympy/integrals/tests/test_integrals.py#L1059) . I don't know why.

jksuom commented on sympy/integrals/integrals.py:
```diff
@@ -519,12 +520,14 @@ def try_meijerg(function, xab):
                     else:
                         try:
-                            function = antideriv._eval_interval(x, a, b)
+                            if b is S.Infinity:
```
This test and the following lines are not necessary as the limit is taken into account in ``_eval_interval``. They seem to change the result in a way that makes some tests fail.

jksuom commented:
The failures seem to be caused by the lines containing ``c``. You should remove those and then test locally with bin/test (at least bin/test stats).

shubhamtibra commented:
What should i use instead of `c`?

jksuom commented:
The issue is not the symbol name. All those lines can be removed. It is not necessary to consider limits at this point.

shubhamtibra committed:
Removed Make.args

shubhamtibra commented:
I did that but now the integral gives `nan`, are we supposed leave the bug then?

jksuom commented:
Hmm.. That is odd. I get this with your latest PR.
```
In [1]: integrate(1/(1+exp(x)),(x,0,oo))
Out[1]: log(2)
```
Could there be a difference in our systems.

shubhamtibra commented:
Now this test is failing `assert integrate(f + exp(z), (z, 2, 3)) == integral_f - exp(2) + exp(3)`.

jksuom commented:
That is true. But that is just an inconvenience, not a real bug. The assert test can be modified, though I may have to look more closely into the associated issue before doing that.

jksuom commented:
The lines concerned with Add were added in commit 6fe73a1f0b0c7d49d0e5c197ec8265a0f428c265 in order to solve issue #2708, but its impact on integrals like the one in this issue were not considered. It is probably best to add ``@XFAIL`` above the test until issue #2708 is cleared up.

shubhamtibra committed:
`def test_issue_2708()` is expected to fail now

[link] (https://github.com/sympy/sympy/blob/master/sympy/integrals/tests/test_integrals.py#L1059)

shubhamtibra commented:
I am still curious why those lines having `c` were causing that error?

jksuom commented:
Some integrals cannot be evaluated and raise a NotImplementedError exception. In that case the unevaluated integral containing the local variable ``c`` leaks out via the except statement skipping the limit line and corrupts the result.

jksuom commented:
Thanks! I'll merge this in a few hours if there are no objections.

shubhamtibra commented:
Thanks for all the help.
